### PR TITLE
PUBDEV-4003: add api tests for python data manipulation module. Add a…

### PR DESCRIPTION
**Pull request by wendycwong**
1. Added python api tests for python commands found in  module Data Manipulation.

2. Changed API tests for cluster and removed the try/except unless we are interested in examining the error message.

3. Changed python documentation on data manipulation commands.  Some of them will give an error if applied to multi-column frames.  However, you only hit the error if you want to return the return H2O Frame.  Otherwise, everything will be fine even though you are getting junk.

### CODE REVIEW

**Comment by wendycwong** [comment]
@h2o-ops please test!

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 1]
```diff
@@ -2818,3 +2820,75 @@ def cannaryHDFSTest(hdfs_name_node, file_name):
...
+    if n>=2:
+        return n*factorial(n-1)
+
+def cumsum(n):
+    """
+    Defined my own cumsum
+
+    :param n:
+    :return:
+    """
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: this is just `n*(n+1)/2`

Comment (wendycwong) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: Done.

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 2]
```diff
@@ -2885,3 +2887,75 @@ def cannaryHDFSTest(hdfs_name_node, file_name):
             return True
         else:       # exception is caused by other reasons.
             return False
+
+def assert_corret_frame_operation(h2oFrame, h2oNewFrame, operString):
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: rename h2oNewFrame -> h2oResultsFrame
h2oFrame -> sourceFrame

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 3]
```diff
@@ -2885,3 +2887,75 @@ def cannaryHDFSTest(hdfs_name_node, file_name):
...
+        'ceil', 'cos', 'cosh', 'cospi', 'cumprod', 'cumsum', 'digamma', 'exp', 'expm1', 'floor', 'round',
+        'sin', 'sign', 'round', 'sinh', 'tan', 'tanh'
+    :return: None.
+    """
+    validStrings = ['acos', 'acosh', 'asin', 'asinh', 'atan', 'atanh', 'ceil', 'cos', 'cosh',
+                     'exp', 'floor', 'gamma', 'lgamma', 'log', 'log10', 'sin', 'sinh',
+                    'sqrt', 'tan', 'tanh', 'trigamma', 'expm1']
+    npValidStrings = ['log2', 'sign']
+    nativeStrings = ['round', 'abs', 'cumsum']
+    multpi = ['cospi', 'sinpi', 'tanpi']
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: ```assert_is_type(operString, validStrings + npValidStrings + nativeStrings + multpi)```

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 4]
```diff
@@ -2885,3 +2887,75 @@ def cannaryHDFSTest(hdfs_name_node, file_name):
...
+    elif operString in multpi:
+        stringOperations = 'result_val = math.'+operString.split('p')[0]+'(h2oFrame[row_ind, col_ind]*math.pi)'
+    else:
+        assert False, operString+" is not a valid command."
+
+    for col_ind in range(h2oFrame.ncols):
+        for row_ind in range(h2oFrame.nrows):
+            exec(stringOperations)
+
+            if abs(h2oNewFrame[row_ind, col_ind]-result_val) > 1e-6:
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: ```assert abs(h2oNewFrame[row_ind, col_ind]-result_val) <= 1e-6, "..."```

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 5]
```diff
@@ -2885,3 +2887,75 @@ def cannaryHDFSTest(hdfs_name_node, file_name):
...
+    elif operString in npValidStrings:
+        stringOperations = 'result_val = np.'+operString+'(h2oFrame[row_ind, col_ind])'
+    elif operString in multpi:
+        stringOperations = 'result_val = math.'+operString.split('p')[0]+'(h2oFrame[row_ind, col_ind]*math.pi)'
+    else:
+        assert False, operString+" is not a valid command."
+
+    for col_ind in range(h2oFrame.ncols):
+        for row_ind in range(h2oFrame.nrows):
+            exec(stringOperations)
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: ```result_val = result_comp(dataFrame[row_ind, col_ind])```

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 6]
```diff
@@ -2885,3 +2887,75 @@ def cannaryHDFSTest(hdfs_name_node, file_name):
...
+        return n*factorial(n-1)
+
+def cumsum(n):
+    """
+    Defined my own cumsum.  Will generate sum from 1 to n.
+
+    :param n:
+    :return:
+    """
+    if n>=0:
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: ```assert n >= 0```

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import h2o
+from h2o.utils.typechecks import assert_is_type
+from h2o.frame import H2OFrame
+
+def h2o_H2OFrame():
+    """
+    Python API test: h2o.frame.H2OFrame(python_obj=None, destination_frame=None, header=0, separator=', ',
+    column_names=None, column_types=None, na_strings=None)
+    """
+    python_lists = [[1,2,3],[4,5,6],["a","b","c"], [1,0,1]]
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame.py]: transpose the list

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_abs.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
+from __future__ import print_function
+import sys
+sys.path.insert(1,"../../../")
+from tests import pyunit_utils
+import h2o
+import numpy as np
+from h2o.utils.typechecks import assert_is_type
+from h2o.frame import H2OFrame
+
+def h2o_H2OFrame_abs():
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_abs.py]: Combine all math ops into a single test (possibly other combinable tests too)

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_all.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+def h2o_H2OFrame_all():
+    """
+    Python API test: h2o.frame.H2OFrame.all()
+    """
+    python_lists = [[1,1,1,1], [1,1,1,1]]
+    h2oframe = h2o.H2OFrame(python_obj=python_lists, na_strings=['NA'])
+    h2oframe[0]=h2oframe[0]>0.0
+    h2oframe[1]=h2oframe[1]>0.0
+    h2oframe[2]=h2oframe[2]>0.0
+    h2oframe[3]=h2oframe[3]>0.0
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_all.py]: noops

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_any.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import sys
+sys.path.insert(1,"../../../")
+import h2o
+from tests import pyunit_utils
+
+def h2o_H2OFrame_any():
+    """
+    Python API test: h2o.frame.H2OFrame.any()
+    """
+    python_lists = [[-1,1,-1,-1], [-1,-1,-1,-1]]
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_any.py]: -1 => 0

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_any_na_rm.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+import h2o
+from tests import pyunit_utils
+
+def h2o_H2OFrame_any_rm_na():
+    """
+    Python API test: h2o.frame.H2OFrame.any_na_rm()
+    """
+    python_lists = [[0,1,0,0], [0,0,0,0]]
+    h2oframe = h2o.H2OFrame(python_obj=python_lists, na_strings=['NA'])
+    assert h2oframe.any(), "h2o.H2OFrame.any_rm_na() command is not working."
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_any_na_rm.py]: any_rm

Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_any_na_rm.py]: also: no NAs in the frame

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_col_names.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+def h2o_H2OFrame_col_names():
+    """
+    Python API test: h2o.frame.H2OFrame.col_names()
+
+    Copied from pyunit_colnames.py
+    """
+    iris_wheader = h2o.import_file(pyunit_utils.locate("smalldata/iris/iris_wheader.csv"))
+    expected_names = ["sepal_len", "sepal_wid", "petal_len", "petal_wid", "class"]
+    assert iris_wheader.col_names == expected_names, \
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_col_names.py]: ```assert iris_wheader.col_names == expected_names == iris_wheader.columns```
and then next test not needed

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_columns_by_type.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    """
+    fr = h2o.import_file(pyunit_utils.locate("smalldata/jira/filter_type.csv"))
+
+    #Positive case. Look for all coltypes
+    num_type = fr.columns_by_type() #numeric by default
+    cat_type = fr.columns_by_type(coltype="categorical")
+    str_type = fr.columns_by_type(coltype="string")
+    time_type = fr.columns_by_type(coltype="time")
+    uuid_type = fr.columns_by_type(coltype="uuid")
+    bad_type = fr.columns_by_type(coltype="bad")
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_columns_by_type.py]: see if can be merged with pyunit_colnames_by_type

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_countmatches.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+def h2o_H2OFrame_countmatches():
+    """
+    Python API test: h2o.frame.H2OFrame.countmatches(pattern)
+
+    Copied from pyunit_countmatches.py
+    """
+    prostate_frame = h2o.import_file(path=pyunit_utils.locate("smalldata/prostate/prostate_cat.csv"))
+    leftover = prostate_frame["DPROS"].cbind(prostate_frame["CAPSULE"]).cbind(prostate_frame["DCAPS"])
+    leftover.ascharacter()  # change to string
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_countmatches.py]: noop

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 15]
```diff
@@ -2885,3 +2887,75 @@ def cannaryHDFSTest(hdfs_name_node, file_name):
...
+    Defined my own factorial just in case using python2.5 or less.
+
+    :param n:
+    :return:
+    """
+    if n>0 and n<2:
+        return 1
+    if n>=2:
+        return n*factorial(n-1)
+
```
Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]:
```
def cumop(items, op):
    res = [None] * len(items)
    for i, item in enumerate(items):
        res[i] = op(res[i-1], item) if i > 0 else item
    return res
```

Comment (st-pasha) [inline h2o-py/tests/pyunit_utils/utilsPY.py]:
```
cumprod = cumop(numbers, lambda x, y: x*y)
cumsum = cumop(numbers, sum)
cummin = cumop(numbers, min)
cummax = cumop(numbers, max)
```

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_cummax.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+import sys
+sys.path.insert(1,"../../../")
+from tests import pyunit_utils
+import h2o
+import numpy as np
+from h2o.utils.typechecks import assert_is_type
+from h2o.frame import H2OFrame
+
+
+def h2o_H2OFrame_cummax():
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_cummax.py]: merge with other math ops using `cumop`

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_day.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
+from __future__ import print_function
+import sys
+sys.path.insert(1,"../../../")
+from tests import pyunit_utils
+import h2o
+from h2o.utils.typechecks import assert_is_type
+from h2o.frame import H2OFrame
+
+def h2o_H2OFrame_day():
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_day.py]: see if all .day, .month, .year, .week, .dayOfWeek, .hour, .minute, .sec, .msec can be combined into a single test, checking against python (or pandas) standard datetime functionality

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_dim.py [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+sys.path.insert(1,"../../../")
+import h2o
+from tests import pyunit_utils
+import numpy as np
+from random import randrange
+
+def h2o_H2OFrame_dim():
+    """
+    Python API test: h2o.frame.H2OFrame.dim
+    """
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_dim.py]: merge with pyunit_h2oH2OfFrame_shape

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_filter_na_cols.py [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+def h2o_H2OFrame_filter_na_cols():
+    """
+    Python API test: h2o.frame.H2OFrame.filter_na_cols(frac=0.2)
+
+    copied from pyunit_filter_nacols.py
+    """
+    fr = h2o.import_file(pyunit_utils.locate("smalldata/logreg/prostate.csv"))
+    include_cols = fr.filter_na_cols()  # should be all columns
+    assert_is_type(include_cols, list)
+    assert include_cols==range(fr.ncol), "h2o.H2OFrame.filter_na_cols() command is not working."
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_filter_na_cols.py]: list(range(fr.ncol))

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_flatten.py [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+
+def h2o_H2OFrame_flatten():
+    """
+    Python API test: h2o.frame.H2OFrame.flatten()
+
+    copied from pyunit_entropy.py
+    """
+    frame = h2o.H2OFrame.from_python(["redrum"])
+    g = frame.flatten()
+    assert g==frame[0,0], "h2o.H2OFrame.flatten() command is not working."
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_flatten.py]: g == "redrum"

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_getrow.py [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+from h2o.exceptions import H2OValueError
+from h2o.utils.typechecks import assert_is_type
+
+
+def h2o_H2OFrame_getrow():
+    """
+    Python API test: h2o.frame.H2OFrame.getrow()
+    """
+    try:
+        numRow = randrange(1,3)
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_getrow.py]: numRow = 1

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_getrow.py [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+        numCol = randrange(1,10)
+        python_lists = np.random.uniform(-1,1, (numRow, numCol))
+        h2oframe = h2o.H2OFrame(python_obj=python_lists)
+        onerow = h2oframe.getrow()
+        assert_is_type(onerow, list)    # check return type
+        # check and make sure random picked elements agree
+        colInd = randrange(0, h2oframe.ncol)
+        assert abs(h2oframe[0, colInd]-onerow[colInd]) < 1e-6, "h2o.H2OFrame.getrow() command is not working."
+    except H2OValueError:   # expect this error when number of row > 1
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_getrow.py]: remove the try-except

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_ischaracter_DEPRECATED.py [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+import sys
+sys.path.insert(1,"../../../")
+import h2o
+from tests import pyunit_utils
+from random import randrange
+from h2o.utils.typechecks import assert_is_type
+import numpy as np
+
+
+def h2o_H2OFrame_ischaracter():
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_ischaracter_DEPRECATED.py]: combine with isstring()

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_lstrip.py [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+def h2o_H2OFrame_lstrip():
+    """
+    Python API test: h2o.frame.H2OFrame.lstrip(set='')
+
+    Copied from runit_lstrip.R
+    """
+    iris = h2o.import_file(path=pyunit_utils.locate("smalldata/iris/iris.csv"))
+    iris["C5"] = iris["C5"].lstrip("Iris-")
+    newNames = iris["C5"].levels()[0]
+    newStrip = ["etosa", "versicolor", "virginica"]
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_lstrip.py]: "setosa" ?

File: h2o-py/h2o/frame.py [hunk 25]
```diff
@@ -1053,7 +1051,7 @@ def as_date(self, format):
         Convert the frame (containing strings / categoricals) into the ``date`` format.
 
         :param str format: the format string (e.g. "YYYY-mm-dd")
```
Comment (wendycwong) [inline h2o-py/h2o/frame.py]: Fixed the example.  Good catch.

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 2 above)
Comment (wendycwong) [inline h2o-py/tests/pyunit_utils/utilsPY.py]: Done

**Comment by navdeep-G** [comment]
@wendycwong can you fix the conflicts for this pr please?

**Comment by jakubhava** [comment]
@wendycwong can you please have a look on @st-pasha comments and resolve the conflicts ? I'm removing the 4release label as this PR is not ready yet

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 3 above)
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 5 above)
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 6 above)
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 4 above)
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 5 above)
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 26]
```diff
@@ -2885,3 +2887,75 @@ def cannaryHDFSTest(hdfs_name_node, file_name):
...
+    elif operString == 'signif':
+        stringOperations = 'result_val = round(h2oFrame[row_ind, col_ind], 7)'
+    elif operString == 'trigamma':
+        stringOperations = 'result_val = scipy.special.polygamma(1,h2oFrame[row_ind, col_ind])'
+    elif operString == 'digamma':
+        stringOperations = 'result_val = scipy.special.polygamma(0,h2oFrame[row_ind, col_ind])'
+    elif operString=='cumprod':
+        stringOperations = 'result_val = factorial(h2oFrame[row_ind, col_ind])'
+    elif operString in validStrings:
+        stringOperations = 'result_val = math.'+operString+'(h2oFrame[row_ind, col_ind])'
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 27]
```diff
@@ -2885,3 +2887,75 @@ def cannaryHDFSTest(hdfs_name_node, file_name):
...
+    elif operString in nativeStrings:
+        stringOperations = 'result_val = '+operString+'(h2oFrame[row_ind, col_ind])'
+    elif operString in npValidStrings:
+        stringOperations = 'result_val = np.'+operString+'(h2oFrame[row_ind, col_ind])'
+    elif operString in multpi:
+        stringOperations = 'result_val = math.'+operString.split('p')[0]+'(h2oFrame[row_ind, col_ind]*math.pi)'
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_abs.py (hunk 8 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_abs.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame.py (hunk 7 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_all.py (hunk 9 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_all.py]: Fixed

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_any.py (hunk 10 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_any.py]: fixed.

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_any_na_rm.py (hunk 11 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_any_na_rm.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_col_names.py (hunk 12 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_col_names.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_columns_by_type.py (hunk 13 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_columns_by_type.py]: There is no such command...

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_countmatches.py (hunk 14 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_countmatches.py]: fixed

File: h2o-py/tests/pyunit_utils/utilsPY.py (hunk 15 above)
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: Done.

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_cummax.py (hunk 16 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_cummax.py]: done

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 28]
```diff
@@ -2885,3 +2887,75 @@ def cannaryHDFSTest(hdfs_name_node, file_name):
...
+    if operString == "log1p":
+        stringOperations = 'result_val=math.log(h2oFrame[row_ind, col_ind]+1)'
+    elif operString == 'signif':
+        stringOperations = 'result_val = round(h2oFrame[row_ind, col_ind], 7)'
+    elif operString == 'trigamma':
+        stringOperations = 'result_val = scipy.special.polygamma(1,h2oFrame[row_ind, col_ind])'
+    elif operString == 'digamma':
+        stringOperations = 'result_val = scipy.special.polygamma(0,h2oFrame[row_ind, col_ind])'
+    elif operString=='cumprod':
+        stringOperations = 'result_val = factorial(h2oFrame[row_ind, col_ind])'
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_day.py (hunk 17 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_day.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_dim.py (hunk 18 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_dim.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_filter_na_cols.py (hunk 19 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_filter_na_cols.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_flatten.py (hunk 20 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_flatten.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_getrow.py (hunk 21 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_getrow.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_getrow.py (hunk 22 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_getrow.py]: Done

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_lstrip.py (hunk 24 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_lstrip.py]: Nope, should be etosa

File: h2o-py/tests/testdir_apis/H2O_Module/pyunit_h2ocluster.py [hunk 29]
```diff
@@ -8,11 +8,7 @@ def h2ocluster():
     """
     Python API test: h2o.cluster()
     """
-
-    try:
-        h2o.cluster()
-    except Exception as e:
-        assert False, "h2o.cluster() command not is working."
+    h2o.cluster()
```
Comment (st-pasha) [inline h2o-py/tests/testdir_apis/H2O_Module/pyunit_h2ocluster.py]: If a function is supposed to return nothing, you can test it as:
```
ret = h2o.cluster()
assert ret is None
```

Comment (st-pasha) [reply h2o-py/tests/testdir_apis/H2O_Module/pyunit_h2ocluster.py]: (search for comments "no return type" in other places too)

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_lstrip.py (hunk 24 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_lstrip.py]: I agree with you.  Let me add it.

File: h2o-py/tests/pyunit_utils/utilsPY.py [hunk 30]
```diff
@@ -3082,3 +3087,85 @@ def check_sorted_2_columns(frame1, sorted_column_indices, prob=0.5, ascending=[T
...
+
+def cumsum(n):
+    """
+    Defined my own cumsum.  Will generate sum from 1 to n.
+
+    :param n:
+    :return:
+    """
+    assert n>=0, "n cannot be negative"
+    return n*(n+1)/2
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-py/tests/pyunit_utils/utilsPY.py]: Deleted it.  Don't think it is used.

File: h2o-py/tests/testdir_apis/H2O_Module/pyunit_h2ocluster.py (hunk 29 above)
Comment (wendycwong) [reply h2o-py/tests/testdir_apis/H2O_Module/pyunit_h2ocluster.py]: Done.  Gone through old tests and added them.

**Comment by wendycwong** [comment]
I incorporated all the new comments.  Thanks, Pasha.

**Comment by st-pasha** [comment]
Thanks, Wendy. Now we just need to appease the gods of Jenkins that make the PRs succeed -- and you're good to go ;)

**Pull request merged**